csp_billing_adapter/utils.py

```diff
@@ -133,3 +136,127 @@ def get_prev_bill_time(
     )
 
     return bill_time
+
+
+def retry_on_exception(
+    func_call: partial,
+    exceptions=Exception,
+    retry_count: int = 3,
+    retry_delay: float = 1,
+    delay_factor: float = 1,
+    logger: logging.Logger = None,
+    func_name: str = None
+):
+    """
+    Retry the provided function call that has been encapsulated using
+    functools.partial as 'func_call', catching the given 'exceptions',
+    up to 'retry_count' times, delaying by 'retry_delay' between tries,
+    multiplied by 'delay_factor' after each attempt, logging retry
+    status messages via 'log' if provided.
+
+    :param func_call:
+        The encapsulated function call that will be retried.
+    :param exceptions:
+        The exception, or tuple of exceptions, to catch and retry the
+        func_call for.
+    :param retry_count:
+        The number of retries to attempt, after an initial failure,
+        before raising any failure.
+    :param retry_delay:
+        The initial number of seconds to wait before retrying the
+        func_call; will be multiplied after each attempt by the
+        delay_factor. Can be a integer or float that is acceptable
+        to time.sleep().
+    :param delay_factor:
+        The factory by which retry_delay will be multiplied after
+        each retry attempt.
+    :param logger:
+        An optional Logger that, if specified, will be used to log
+        retry status messages.
+    :param func_name:
+        An optional function name that will be used for logging
+        purposes.
+
+    :return:
+        The result of calling func_call.
+    """
+
+    if logger is None:
+        # if no specific logger was provided use default log
+        logger = log
+
+    if func_name is None:
+        # if no function name is provided try using the name of
+        # func_call.func, or failing that, 'unnamed_function'
+        func_name = getattr(
```

Comment:
Why wouldn't a function name be provided?

Reply:
So for most functions, the dynamic lookup of func_call.func.__name__ works, but for hook calls this attribute doesn't exist, so I plumbed in being able to provide a func_name explicitly for use with the hook calls, but for most other cases the dynamic retrieval will work.

Reply:
We potentially could leverage retry_on_exception() in any of the plugins by importing it from csp_billing_adapter.utils, and in such cases, when not dealing with hook calls, passing a func_name wouldn't be required anymore.

Reply:
And we could implement a decorator based upon retry_on_exception() if that would be beneficial...

Reply:
Seems good for now. :+1: